### PR TITLE
Add shadows plugin

### PR DESCRIPTION
**Pull request by tomas7770**
# Description

Adds the base interface plugin for shadows, which registers shadow caster components.

## Checklist

- [x] Self-review changes.
- [ ] Evaluate impact on the documentation.
- [ ] Ensure test coverage.
- [ ] Write new samples.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: engine/include/cubos/engine/render/shadows/caster.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/// @file
+/// @brief Component @ref cubos::engine::ShadowCaster.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/render/shadows/caster.hpp:2]:
```suggestion
/// @brief Struct @ref cubos::engine::ShadowCaster.
```

This isn't meant to be used as a component, right?

File: engine/include/cubos/engine/render/shadows/caster.hpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+#include <glm/vec2.hpp>
+
+#include <cubos/core/reflection/reflect.hpp>
+
+#include <cubos/engine/api.hpp>
+
+namespace cubos::engine
+{
+    /// @brief Component which contains the options common to all shadow casters.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/render/shadows/caster.hpp:15]:
```suggestion
    /// @brief Struct which contains the options common to all shadow casters.
```

File: engine/include/cubos/engine/render/shadows/caster.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    /// @ingroup render-shadows-plugin
+    struct CUBOS_ENGINE_API ShadowCaster
+    {
+        CUBOS_REFLECT;
+
+        float bias;               ///< Shadow bias.
+        float blurRadius;         ///< Shadow blur radius.
+        glm::uvec2 shadowMapSize; ///< Size of the shadow map, in pixels.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/render/shadows/caster.hpp:23]: Hm, from our discussion yesterday regarding the shadow atlas, ig this no longer makes sense.
Maybe just omit it for now? Maybe later on we can assign some kind of 'priority', where casters with higher priorities get bigger texture slots or something.

File: engine/include/cubos/engine/render/shadows/directional_caster.hpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace cubos::engine
+{
+    /// @brief Component which enables shadow casting on a directional light.
+    /// @ingroup render-shadows-plugin
+    struct CUBOS_ENGINE_API DirectionalShadowCaster
+    {
+        CUBOS_REFLECT;
+
+        cubos::engine::ShadowCaster options;
```
Comment (RiscadoA) [inline engine/include/cubos/engine/render/shadows/directional_caster.hpp:22]: Hm I think we could choose a better field name, as the other fields are also options.
Maybe `base` or `common`?

File: engine/include/cubos/engine/render/shadows/point_caster.hpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace cubos::engine
+{
+    /// @brief Component which enables shadow casting on a point light.
+    /// @ingroup render-shadows-plugin
+    struct CUBOS_ENGINE_API PointShadowCaster
+    {
+        CUBOS_REFLECT;
+
+        cubos::engine::ShadowCaster options;
```
Comment (RiscadoA) [inline engine/include/cubos/engine/render/shadows/point_caster.hpp:20]: Same as above

File: engine/include/cubos/engine/render/shadows/spot_caster.hpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace cubos::engine
+{
+    /// @brief Component which enables shadow casting on a spot light.
+    /// @ingroup render-shadows-plugin
+    struct CUBOS_ENGINE_API SpotShadowCaster
+    {
+        CUBOS_REFLECT;
+
+        cubos::engine::ShadowCaster options;
```
Comment (RiscadoA) [inline engine/include/cubos/engine/render/shadows/spot_caster.hpp:20]: Same as above

**Review by RiscadoA — APPROVED** [review]
Mostly LGTM! Thanks for working on this!

File: engine/include/cubos/engine/render/shadows/caster.hpp (hunk 2 above)
Comment (DiogoMendonc-a) [inline engine/include/cubos/engine/render/shadows/caster.hpp:15]:
```suggestion
    /// @brief Component which contains the settings common to all shadow casters.
```
Sounds better imo, but it's up to you.

**Review by DiogoMendonc-a — APPROVED** [review]
👍

File: engine/include/cubos/engine/render/shadows/directional_caster.hpp (hunk 4 above)
Comment (fallenatlas) [reply engine/include/cubos/engine/render/shadows/directional_caster.hpp:22]: I agree with this, `base` or `baseSettings` makes the most sense to me

**Review by fallenatlas — APPROVED** [review]
LGTM!

File: engine/include/cubos/engine/render/shadows/directional_caster.hpp (hunk 4 above)
Comment (tomas7770) [reply engine/include/cubos/engine/render/shadows/directional_caster.hpp:22]: `baseSettings` sounds the most clear to me.

File: engine/include/cubos/engine/render/shadows/caster.hpp (hunk 3 above)
Comment (tomas7770) [reply engine/include/cubos/engine/render/shadows/caster.hpp:23]: Yeah, you're right, it makes more sense to think of these details when the shadow atlas is implemented.

**Pull request merged**